### ISSUE TITLE
Make `isEmpty` a type guard

### ISSUE THREAD

**Issue by yokuze** [silvermine/toolbox#16]
Currently, `isEmpty` just returns a `boolean`. It would be much more useful if it were a type-guard that asserted that its parameter was one of the "empty" object types its checking for.

**Comment by crgwbr** [comment]
@yokuze This is easy for `null`, `undefined`, `array`, and `string` types. But `isEmpty` currently also works for the `IArguments` type, which isn't generic. So, there is no type for empty arguments vs. non-empty arguments. So… thoughts?

**Comment by yokuze** [comment]
I don't think we need to differentiate between empty vs non-empty `IArguments` in the type guard assertion. It would be nifty if there were an existing way to do so, but I don't think it'll be that helpful to us to make that distinction.

By far, the most frequent pain point that we are trying to solve by making this function a type guard is something like this:

```typescript
function doSomething(withThis: string | undefined): void {
   if (isEmpty(withThis)) {

   } else {
      // TypeScript should at least know here that `withThis` not `undefined`
   }
}
```

Anything beyond that is extra credit IMO.

That said, I may be missing something, so let me know if there's a use case you think we should support by differentiating between empty vs non-empty `IArguments`. If there is, the only option I can think of right now would be to extend the `IArguments` type like:

```typescript
interface IEmptyArguments extends IArguments {
   length: 0;
}
```
